### ISSUE TITLE
Argument to class must be a base class

### ISSUE THREAD

**Issue by brianferri** [microsoft/pyright#9490]
> [!IMPORTANT]  
> The app works completely fine in runtime and produces no errors, it's just the linter giving the [TITLE] error

## Environment data

-   Pylance version: v2024.11.2
-   OS and version: Darwin Kernel Version 24.1.0: arm64 (Sequoia 15.1.1 (24B91))
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.11.10 (Conda)

### minimal repro

```python
from enum import Enum
from typing import Generic, TypeVar

T=TypeVar("T")

class A(Generic[T], Enum): pass

class B(A[int]): pass # Argument to class must be a base class Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

### use case

```py
from enum import Enum
from pydantic import BaseModel
from typing import Generic, TypeVar
from abc import ABC, abstractmethod


class Workflow(BaseModel, ABC):
    @property
    @abstractmethod
    def chat_interface(self):
        raise NotImplementedError(
            "Subclasses must implement the 'chat_interface' property")


TEnum = TypeVar("TEnum")


class NamedEnum(Generic[TEnum], Enum):
    """
    Effectively swaps the key with the value in an enum
    """

    def __init__(self, *args, **kwargs):
        self._old: TEnum = self._value_
        self._value_ = self._name_

    @classmethod
    def _missing_(cls, value: str):
        for member in cls:
            if member.name == value:
                return member
        return None

    @property
    def orig(self): return self._old


class ModelFlow(NamedEnum[Workflow]): # Argument to class must be a base class Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
    """
    Dynamically generated Enum from the `Workflow` subclasses.
    """
    @staticmethod
    def _generate_enum():
        subclasses = Workflow.__subclasses__()
        return {
            subclass.__name__: subclass
            for subclass in subclasses
        }

    locals().update(_generate_enum())
```

So that I can access `.orig` from a ModelFlow and use it as if it were a `Workflow`

## Expected behavior

That it not error and that the `.orig(...)` type be a `Workflow`

## Actual behavior

`.orig` is `(property) orig: (self: Self@NamedEnum[TEnum@NamedEnum]) -> TEnum@NamedEnum`

## Logs

```
2024-11-21 18:59:22.850 [info] (23211) Attempting to resolve using local imports: Enum
2024-11-21 18:59:22.851 [info] (23211) Attempting to resolve using root path 'file:///projects/src/langchain/workflows'
2024-11-21 18:59:22.851 [info] (23211) Attempting to resolve using root path 'file:///projects/src/langchain'
2024-11-21 18:59:22.851 [info] (23211) Attempting to resolve using root path 'file:///projects/src'
2024-11-21 18:59:22.931 [info] (23211) [FG] parsing: file:///projects/.conda/lib/python3.11/site-packages/pydantic/_internal/_decorators.py [fs read 1ms] (8ms)
2024-11-21 18:59:22.933 [info] (23211) [FG] binding: file:///projects/.conda/lib/python3.11/site-packages/pydantic/_internal/_decorators.py (1ms)
2024-11-21 18:59:23.243 [info] (23211) [FG] parsing: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:23.243 [info] (23211) [FG] binding: file:///projects/src/langchain/workflows/__init__.py (0ms)
2024-11-21 18:59:23.344 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211959945 at file:///projects/src/langchain/workflows/__init__.py ...
2024-11-21 18:59:23.345 [info] (23211) [BG(1)]   parsing: file:///projects/src/langchain/workflows/__init__.py (0ms)
2024-11-21 18:59:23.345 [info] (23211) [BG(1)]   binding: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:23.412 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211959945 at file:///projects/src/langchain/workflows/__init__.py (68ms)
2024-11-21 18:59:23.727 [info] (23211) [FG] parsing: file:///projects/src/langchain/workflows/__init__.py (0ms)
2024-11-21 18:59:23.727 [info] (23211) [FG] binding: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:23.829 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211963345 at file:///projects/src/langchain/workflows/__init__.py ...
2024-11-21 18:59:23.829 [info] (23211) [BG(1)]   parsing: file:///projects/src/langchain/workflows/__init__.py (0ms)
2024-11-21 18:59:23.829 [info] (23211) [BG(1)]   binding: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:23.896 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211963345 at file:///projects/src/langchain/workflows/__init__.py (68ms)
2024-11-21 18:59:24.029 [info] (23211) [BG(1)] indexing: file:///projects/src/langchain/workflows/__init__.py [found 2] (0ms)
2024-11-21 18:59:24.029 [info] (23211) Workspace indexing done: file:///projects/src/langchain/workflows/__init__.py
2024-11-21 18:59:24.030 [info] [Info  - 18:59:24] (23211) Heap stats: total_memory_size=16384MB, total_free_size=125MB, total_heap_size=188MB, used_heap_size=140MB, cross_worker_used_heap_size=140MB, total_physical_size=188MB, total_available_size=3956MB, heap_size_limit=4096MB
2024-11-21 18:59:24.143 [info] (23211) [BG(1)] analyzing: file:///projects/src/api/routers/collections.py ...
2024-11-21 18:59:24.143 [info] (23211) [BG(1)]   checking: file:///projects/src/api/routers/collections.py (114ms)
2024-11-21 18:59:24.143 [info] (23211) [BG(1)] analyzing: file:///projects/src/api/routers/collections.py (114ms)
2024-11-21 18:59:24.145 [info] (23211) [BG(1)] analyzing: file:///projects/src/langchain/workflows/__init__.py ...
2024-11-21 18:59:24.145 [info] (23211) [BG(1)]   checking: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:24.145 [info] (23211) [BG(1)] analyzing: file:///projects/src/langchain/workflows/__init__.py (1ms)
2024-11-21 18:59:24.147 [info] (23211) [BG(1)] analyzing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/builtins.pyi ...
2024-11-21 18:59:24.148 [info] (23211) [BG(1)]   checking: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/builtins.pyi ...
2024-11-21 18:59:24.148 [info] (23211) [BG(1)]     parsing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/_ast.pyi [fs read 1ms] (2ms)
2024-11-21 18:59:24.148 [info] (23211) [BG(1)]     binding: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/_ast.pyi (1ms)
2024-11-21 18:59:24.152 [info] (23211) [BG(1)]     parsing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/sys/__init__.pyi [fs read 0ms] (4ms)
2024-11-21 18:59:24.153 [info] (23211) [BG(1)]     binding: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/sys/__init__.pyi (1ms)
2024-11-21 18:59:24.156 [info] (23211) [BG(1)]     parsing: file:///projects/.conda/lib/python3.11/types.py [fs read 1ms] (3ms)
2024-11-21 18:59:24.157 [info] (23211) [BG(1)]     binding: file:///projects/.conda/lib/python3.11/types.py (1ms)
2024-11-21 18:59:24.159 [info] (23211) [BG(1)]     parsing: file:///projects/.conda/lib/python3.11/collections/abc.py [fs read 0ms] (0ms)
2024-11-21 18:59:24.159 [info] (23211) [BG(1)]     binding: file:///projects/.conda/lib/python3.11/collections/abc.py (1ms)
2024-11-21 18:59:24.165 [info] (23211) [BG(1)]     parsing: file:///projects/.conda/lib/python3.11/io.py [fs read 0ms] (6ms)
2024-11-21 18:59:24.167 [info] (23211) [BG(1)]     binding: file:///projects/.conda/lib/python3.11/io.py (1ms)
2024-11-21 18:59:24.183 [info] (23211) [BG(1)]     parsing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/io.pyi [fs read 0ms] (12ms)
2024-11-21 18:59:24.188 [info] (23211) [BG(1)]     binding: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/io.pyi (5ms)
2024-11-21 18:59:24.194 [info] (23211) [BG(1)]     parsing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/_io.pyi [fs read 3ms] (10ms)
2024-11-21 18:59:24.195 [info] (23211) [BG(1)]     binding: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/_io.pyi (1ms)
2024-11-21 18:59:24.221 [info] (23211) [BG(1)]     parsing: file:///projects/.conda/lib/python3.11/site-packages/typing_extensions.py [fs read 2ms] (25ms)
2024-11-21 18:59:24.226 [info] (23211) [BG(1)]     binding: file:///projects/.conda/lib/python3.11/site-packages/typing_extensions.py (5ms)
2024-11-21 18:59:24.343 [info] (23211) [BG(1)]     parsing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/ast.pyi [fs read 2ms] (9ms)
2024-11-21 18:59:24.346 [info] (23211) [BG(1)]     binding: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/ast.pyi (3ms)
2024-11-21 18:59:24.414 [info] (23211) [BG(1)]   checking: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (269ms)
2024-11-21 18:59:24.415 [info] (23211) [BG(1)] analyzing: file:///Users/brianferri/.vscode/extensions/ms-python.vscode-pylance-2024.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (269ms)
2024-11-21 18:59:24.419 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211963829 at file:///projects/src/langchain/workflows/__init__.py (0ms)
2024-11-21 18:59:24.419 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211847089 at file:///projects/src/langchain/workflows/Workflow.py (0ms)
2024-11-21 18:59:24.419 [info] (23211) [BG(1)] SemanticTokens delta previousResultId:1732211959946 at file:///projects/src/api/routers/collections.py (0ms)

```

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

Contrary to what the bug report says, if you attempt to run the code above, you will see a runtime exception. That's because the `EnumMeta` metaclass defines a custom `__getitem__` method which is invoked for the expression `A[int]`.
